lib/AST/ShouldPrintForParseableInterface.cpp

```diff
@@ -83,12 +92,63 @@ class FindReferencedNonPublicTypeDecls: public ASTWalker {
   FindTypeDecls findTypeDecls;
 public:
   FindReferencedNonPublicTypeDecls(TypeDecls &foundDecls):
-    findTypeDecls(foundDecls) {}
+    findTypeDecls(foundDecls, workList) {}
+
+  void search(ModuleDecl *module) {
+    SmallVector<Decl *, 8> topLevelDecls;
+    module->getTopLevelDecls(topLevelDecls);
+    for (auto decl : topLevelDecls) {
+      auto vd = dyn_cast<ValueDecl>(decl);
+      if (!vd || !isPublicOrUsableFromInline(vd))
+        continue;
+      workList.push_back(vd);
+    }
+
+    while (!workList.empty()) {
+      Decl *d = workList.pop_back_val();
+      d->walk(*this);
+    }
+  }
+
+  void visitAbstractStorageDecl(AbstractStorageDecl *decl) {
+    if (!contributesToParentTypeStorage(decl))
+      return;
+    decl->getInterfaceType().walk(findTypeDecls);
+  }
+
+  void visitEnumElementDecl(EnumElementDecl *decl) {
+    auto params = decl->getParameterList();
+    if (!params) return;
+    for (auto param : params->getArray()) {
+      param->getInterfaceType().walk(findTypeDecls);
+    }
+  }
+
+  void visitProtocolDecl(ProtocolDecl *decl) {
+    for (auto inherited : decl->getInheritedProtocols())
+      inherited->getInterfaceType().walk(findTypeDecls);
+  }
+
   bool walkToDeclPre(Decl *decl) {
-    auto asd = dyn_cast<AbstractStorageDecl>(decl);
-    if (!asd) return true;
-    if (contributesToParentTypeStorage(asd))
-      asd->getInterfaceType().walk(findTypeDecls);
+    if (auto storage = dyn_cast<AbstractStorageDecl>(decl)) {
+      visitAbstractStorageDecl(storage);
+      return false;
+    }
+
+    if (auto enumElement = dyn_cast<EnumElementDecl>(decl)) {
+      visitEnumElementDecl(enumElement);
+      return false;
+    }
+
+    if (auto protocol = dyn_cast<ProtocolDecl>(decl)) {
+      visitProtocolDecl(protocol);
```

Comment:
I don't think the inheritance thing just applies to protocols. We also care about the layout of `@objc` enums, and if we're going to print classes we should probably print their superclass.